File: IntToEnglish.java
//T: O(N)
//S: O(1)
class Solution {
           
 String[] units={"","One","Two","Three","Four","Five","Six","Seven","Eight","Nine","Ten" };
        
String [] tens={"","Eleven","Twelve","Thirteen","Fourteen","Fifteen","Sixteen","Seventeen","Eighteen","Nineteen"} ;  
String [] places={"","Thousand ","Million ","Billion "};    
 
    String[] tees={"","","Twenty","Thirty","Forty","Fifty","Sixty","Seventy","Eighty","Ninety"}; 
    
    public String numberToWords(int num) {
        if(num==0){return "Zero";}
       int feed=num;
        String ans="";
        int pl=0;
        while(feed!=0 ){
            
            String ret=helper(feed%1000);
            
            feed=feed/1000;
            if(!ret.equals("")){
              ans=ret+" "+places[pl]+ans;
                
            }
            pl++;
        }
        
        
        return ans.trim();
    }
    
    public String helper(int num){
        
        if(num<=10){
            
            return units[num];
        }else if(num<20){
            return tens[num%10];
            
        }else if(num<=99){
            if(num%10==0){ return tees[num/10];}
            return tees[num/10] +" "+helper(num%10);
        }else{
            
           if(num%100==0){return units[num/100]+" Hundred";}
            
        return units[num/100]+" Hundred "+helper(num%100);
        }
         
        
    }
    
    
    
}

File: int to english.py
# // Time Complexity : O(1)
# // Space Complexity : O(1)
#  // Did this code successfully run on Leetcode : Yes
# // Any problem you faced while coding this : No


# // Your code here along with comments explaining your approach
# Process numbers in hundreds i.e 3 numbers => divide number by 1000 in each iteration
# use three arrays : under_20, tens, thousands: each will have empty string at start for convenience
# create convertor finction: 
# where if number is <20 we give out result from under_20 array
# if <100 we do n/10 and get the corresponding number from under_20 and call convertor on n%10
# for rest we do n/100 and get the corresponding number from under_20 and call convertor on n%100


class Solution:
    def __init__(self):
        self.below_20 = []
        self.tens = []
        self.thousands = []
        
    def numberToWords(self, num: int) -> str:
        if num == 0: return "Zero"
        #used for 1,2,3 digits
        self.below_20 = ["","One","Two",'Three','Four','Five','Six',"Seven","Eight",'Nine',"Ten","Eleven","Twelve","Thirteen","Fourteen","Fifteen","Sixteen","Seventeen","Eighteen","Nineteen"]
        #used for more than 2 digits
        self.tens = ["","Ten","Twenty","Thirty","Forty","Fifty","Sixty","Seventy","Eighty","Ninety"]
        self.thousands = ["","Thousand","Million","Billion"]
        res = ''
        i = 0
        while num:
            # if num is not 000
            if num % 1000 != 0:
                # call convertor on num%1000
                res = self.converter(num%1000) + self.thousands[i] +" "+ res
            num = num//1000
            i += 1
            
        return res.rstrip()
    
    def converter(self, n):
        r = ''
        if n == 0:
            return ""
        elif n<20:
            return self.below_20[n] +" "
        elif n<100:
            return self.tens[n//10] +" "+ self.converter(n%10)
        else:
            return self.below_20[n//100] + " Hundred "+ self.converter(n%100)
        